Created `platform_designer_standard` sample for CycloneV SoC (#4)
haoyanwa/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/Tutorials/Tools/platform_designer_standard/README.md

Reviewer: don't forget to update register your new code sample with the top-level readme

   2. Add the following IP to your system:

      * Basic Functions > Bridges and Adaptors > Memory Mapped > **JTAG to Avalon Master Bridge Intel® FPGA IP**

      * oneAPI > **add_report_di**

      ![](assets/add-ip-platform-designer.png)

      > **Note**: If you cannot see the oneAPI IP component, refresh Platform Designer by clicking `File` > `Refresh System`
      > ![](assets/refresh-system.png)

   3. Connect the modules as shown:

      ![](assets/module-connections.png)

      Don't forget to export the `irq_add` and `exception_add` signals. The provided top-level RTL file (`add.sv`) uses the generated IP. Following these naming conventions allows the IP to connect to this handwritten RTL.

   4. Save the system by clicking `File` > `Save` and name it `add_kernel_wrapper.qsys`.

   5. Make sure there are no errors in the 'System Messages' panel.

   6. Generate the system so that it can be included in the Intel® Quartus® Prime project by clicking `Generate HDL...`

      ![](assets/generate-hdl.png)

   7. Close Platform Designer. 

>    

Reviewer 2: When I reach this point, I was prompted to add .qip and .sip file manually. 
![Screenshot 2024-09-06 162404](https://github.com/user-attachments/assets/262c11f7-219f-471c-a567-bf4d5ff3beee)

If I did nothing, the next action "Analysis and Elab" will fail with error. 
![Screenshot 2024-09-06 162813](https://github.com/user-attachments/assets/a21c481f-968d-4c6e-a470-29c19212be9f)

The Analysis and Elab would fail and could only proceed after I add .qip file.
![image](https://github.com/user-attachments/assets/28fb18cf-39c0-40d8-99bc-726e45255132)

Reviewer 2: I dont remember this happening when I go thru the platform designer with Quartus Pro.
For people like me who are not familiar with PD, the screenshot below do helps me how to use the navigate to add qip file.
![image](https://github.com/user-attachments/assets/f862bce9-a3ea-43f4-901b-29640b273845)

Author: Will add this step as a note in the README. Thank you for the suggestion!

   1. Open the pin planner using `Assignments` > `Pin Planner` in the main Intel® Quartus® Prime GUI. In the bottom pane of the Pin Planner GUI, you will find a list of signals to assign pins to. Consult the data sheet for your board to choose an appropriate clock input. Double-click the *Location* cell and select a proper clock input. In this project, the `PIN_AF14` was chosen because it supplies a 50MHz clock signal.

   2. Assign pins for the `fpga_led` and `reset_button` signals using the same method. Your final pin assignment should match with the following screenshot.

      *Final pin planner configuration:*

      ![](assets/pins-assignment.png)

      > **Note**: If you cannot see the pin details, click the `All Pins` button in the top left corner of the bottom pane in Pin Planner GUI.
      > 
      > ![](assets/all-pins.png)

      > **Note**: Make sure you choose 'LVDS' for the I/O standard of `i_clk`, the pin location will be automatically populated for `i_clk(n)`.

7. Compile the full design by clicking the 'Start Compilation' button in the Intel® Quartus® Prime GUI.

      ![](assets/start-compilation-quartus.png)

> 

Reviewer 2: I got "Info (293000): Quartus Prime Full Compilation was successful. 0 errors, 380 warnings"

I think this is what we expected so far

8. Copy the generated `add.sof` file to the `system_console` directory.

   Linux:

   ```bash
   $> cp add_quartus/output_files/add.sof system_console
   ```

   Windows:

   ```bash
   > xcopy add_quartus\output_files\add.sof system_console /Y
   ```

> You may also build the SOF using the pre-generated Intel® Qupartus® Prime project in the `add_quartus_sln` directory by executing the included `build_system.tcl` script. This script has been verified against the latest version of Quartus® Prime Standard Edition software available at the time of writing (23.1). The script and pre-generated project may not work with other versions of Quartus® Prime.

Reviewer: ```suggestion
You may also build the SOF using the pre-generated Intel® Quartus® Prime project in the `add_quartus_sln` directory by executing the included `build_system.tcl` script. This script has been verified against the latest version of Quartus® Prime Standard Edition software available at the time of writing (23.1). The script and pre-generated project may not work with other versions of Quartus® Prime.
```

- [Intel® Cyclone® V FPGA and SoC FPGA](https://www.intel.com/content/www/us/en/products/details/fpga/cyclone/v.html) describes the Intel® Cyclone® V FPGA in greater detail.
- [Intel® Quartus® Prime Standard Edition User Guide: Getting Started](https://www.intel.com/content/www/us/en/docs/programmable/683475/current/introduction-to.html) introduces you to the Intel® Quartus® Prime Standard software.
- [Intel® Quartus® Prime Standard Edition User Guide: Platform Designer](https://www.intel.com/content/www/us/en/docs/programmable/683364/current/creating-a-system-with.html) describes the Intel® Platform Designer software.
- [Intel® Quartus® Prime Standard Edition User Guide: Programmer](https://www.intel.com/content/www/us/en/docs/programmable/683528/current/programming-fpga-devices.html) describes the Intel® Quartus® Prime Standard Programmer software.
- [Intel® Quartus® Prime Standard Edition User Guide: Debug Tools](https://www.intel.com/content/www/us/en/docs/programmable/683552/current/system-console-commands.html) describes the console commands enable testing.

> 

Reviewer 2: I did not had a board with me, so I stopped here. All went well except the part where I have to manually add the IP .qip.

Author: Thank you! I tested on the board multiple times, it worked as per the description

# `Platform Designer Standard` Sample

> This example design shows how to use an FPGA IP produced with the Intel® oneAPI DPC++/C++ Compiler with the Intel® Quartus® Prime Standard Edition to target a Cyclone® V FPGA.

Reviewer: you can modify the OG platform designer sample to remove mention of Quartus std since we show it here now.

| Optimized for                     | Description
|:---                               |:---
| OS                                | Ubuntu* 20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10 <br> Windows Server* 2019

> | Hardware                          | This process applies to the [Cyclone® V FPGA and SoC FPGA](https://www.intel.com/content/www/us/en/products/details/fpga/cyclone/v.html)

Reviewer: mention the specific development kit that you target so that your customer knows which board to use, like the OG platform desginer sample

![](assets/csr-output-example-simple.svg)

This example is intended for users interested in creating standalone modules that can be included in Intel® Quartus® Prime projects. It serves as a minimal example, and while it targets a specific board, a user familiar with the Intel® Quartus® Prime Software Suite should be able to easily port this design to other hardware.

## Key Implementation Details

This tutorial is structured with four source code directories.

| Source Directory  | Description
|:--                |:--
| `add_oneapi`      | The source and build scripts needed to compile a simple IP using oneAPI
| `add_quartus_sln` | An example of the project files that are created using the Intel® Quartus® Prime GUI
| `starting_files`  | The minimal source files you need to create a Intel® Quartus® Prime project. Additional project files are created using the Intel® Quartus® Prime GUI.
| `system_console`  | Scripts for controlling the System Console application while you test the design in hardware

### Board-specific Considerations

> This design is intended to work with the Intel® Cyclone® V SoC FPGA. These board-specific configurations are not guaranteed to work with different boards:

Reviewer: ```suggestion
This design is intended to work with the <Terasic blahblahblah developer kit>. These board-specific configurations are not guaranteed to work with different boards:
```

> This practice ensures that your compiler, libraries, and tools are ready for development.
>
> Linux*:
> - For system-wide installations: `. /opt/intel/oneapi/setvars.sh`
> - For private installations: ` . ~/intel/oneapi/setvars.sh`
> - For non-POSIX shells, like csh, use the following command: `bash -c 'source <install-dir>/setvars.sh ; exec csh'`
>
> Windows*:
> - `C:\"Program Files (x86)"\Intel\oneAPI\setvars.bat`
> - Windows PowerShell*, use the following command: `cmd.exe "/K" '"C:\Program Files (x86)\Intel\oneAPI\setvars.bat" && powershell'`
>
> For more information on configuring environment variables, see [Use the setvars Script with Linux* or macOS*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html) or [Use the setvars Script with Windows*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).

> **Follow these steps to compile and test the design:**

Reviewer: ```suggestion
**Follow these steps to compile and test the design:**
```
why did you bold this text? if you bold this, you should also bold the Quartus Pro version so they are consistent

      > **Note**: If you cannot see the oneAPI IP component, refresh Platform Designer by clicking `File` > `Refresh System`
      > ![](assets/refresh-system.png)

   3. Connect the modules as shown:

      ![](assets/module-connections.png)

      Don't forget to export the `irq_add` and `exception_add` signals. The provided top-level RTL file (`add.sv`) uses the generated IP. Following these naming conventions allows the IP to connect to this handwritten RTL.

   4. Save the system by clicking `File` > `Save` and name it `add_kernel_wrapper.qsys`.

   5. Make sure there are no errors in the 'System Messages' panel.

   6. Generate the system so that it can be included in the Intel® Quartus® Prime project by clicking `Generate HDL...`

      ![](assets/generate-hdl.png)

   7. Close Platform Designer. 
   
5. In the Intel® Quartus® Prime window, run Analysis and Elaboration by clicking 'Start Analysis and Elaboration'.

   ![](assets/start-analysis.png)

6. Select pins for the `i_clk` and `reset_button_n` inputs and `fpga_led` output. The JTAG to Avalon® Master Bridge Intel FPGA IP handles the connection between your design and the JTAG pins on your board automatically.

>    1. Open the pin planner using `Assignments` > `Pin Planner` in the main Intel® Quartus® Prime GUI. In the bottom pane in Pin Planner GUI, you will find a list of signals to be assigned a pin. Consult the data sheet for your board to choose an appropriate clock input. Double-click the *Location* cell and select a proper clock input. In this project, the `PIN_AF14` was chosen because it supplies a 50MHz clock signal.

Reviewer: ```suggestion
   1. Open the pin planner using `Assignments` > `Pin Planner` in the main Intel® Quartus® Prime GUI. In the bottom pane of the Pin Planner GUI, you will find a list of signals to assign pins to. Consult the data sheet for your board to choose an appropriate clock input. Double-click the *Location* cell and select a proper clock input. In this project, the `PIN_AF14` was chosen because it supplies a 50MHz clock signal.
```

- [Intel® Quartus® Prime Standard Edition User Guide: Debug Tools](https://www.intel.com/content/www/us/en/docs/programmable/683552/current/system-console-commands.html) describes the console commands enable testing.

## Running the Sample

Use the `test.bat` script in the `system_console` directory to flash the design to your development board, and launch the system console to execute the testing script. During the programming process, the TX LED on the board will illuminate, indicating successful programming. During the test, each data transaction will trigger both the TX and RX LEDs to blink, indicating active communication. The included `.tcl` script in the `system_console` directory demonstrates how to use the System Console to interact with your IP through the JTAG to Avalon® Master Bridge Intel FPGA IP on the FPGA.

To move the design to a different computer for testing, copy the entire `system_console` directory, along with `add.sof`.

See output:

```

> Info: *******************************************************************
> Info: Running Quartus Prime Programmer
>     Info: Version 23.1std.1 Build 993 05/14/2024 SC Standard Edition
>     Info: Copyright (C) 2024  Intel Corporation. All rights reserved.
>     Info: Your use of Intel Corporation's design tools, logic functions
>     Info: and other software and tools, and any partner logic
>     Info: functions, and any output files from any of the foregoing
>     Info: (including device programming or simulation files), and any
>     Info: associated documentation or information are expressly subject
>     Info: to the terms and conditions of the Intel Program License
>     Info: Subscription Agreement, the Intel Quartus Prime License Agreement,
>     Info: the Intel FPGA IP License Agreement, or other applicable license
>     Info: agreement, including, without limitation, that your use is for
>     Info: the sole purpose of programming logic devices manufactured by
>     Info: Intel and sold by Intel or its authorized distributors.  Please
>     Info: refer to the applicable agreement for further details, at
>     Info: https://fpgasoftware.intel.com/eula.
>     Info: Processing started: Mon Aug 26 13:33:34 2024
> Info: Command: quartus_pgm -m jtag -c 1 -o p;add.sof@2
> Info (213045): Using programming cable "DE-SoC [USB-1]"
> Info (213011): Using programming file add.sof with checksum 0x00C70E5A for device 5CSEMA5F31@2
> Info (209060): Started Programmer operation at Mon Aug 26 13:33:36 2024
> Info (209016): Configuring device index 2
> Info (209017): Device 2 contains JTAG ID code 0x02D120DD
> Info (209007): Configuration succeeded -- 1 device(s) configured
> Info (209011): Successfully performed operation(s)
> Info (209061): Ended Programmer operation at Mon Aug 26 13:33:39 2024

Reviewer: ```suggestion
<output from Intel® Quartus® Prime programmer>
```

Author: This is perfect. I also felt the whole log redundant.

# `Platform Designer Standard` Sample

This example design shows how to use an FPGA IP produced with the Intel® oneAPI DPC++/C++ Compiler with the Intel® Quartus® Prime Standard Edition.

| Optimized for                     | Description
|:---                               |:---
| OS                                | Ubuntu* 20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10 <br> Windows Server* 2019

> | Hardware                          | This process applies to the [Cyclone® V SE SoC FPGAs](https://www.intel.com/content/www/us/en/products/details/fpga/cyclone/v/se.html), but the sample Intel® Quartus® Prime Standard Edition project targets the [Terasic DE1-SOC Development Board](https://de1-soc.terasic.com/)

Reviewer: ```suggestion
| Hardware                          | This process applies to the Cyclone® V line of FPGAs, but the sample Intel® Quartus® Prime Standard Edition project targets the [Terasic DE1-SOC Development Board](https://de1-soc.terasic.com/)
```
no need to overconstrain ourselves

### Board-specific Considerations

> This design is intended to work with the [Terasic DE1-SOC Development Board](https://de1-soc.terasic.com/). These board-specific configurations are not guaranteed to work with different boards:

Reviewer: ```suggestion
This design is intended to work with the Terasic DE1-SOC Development Board. These board-specific configurations are not guaranteed to work with different boards:
```
you only need a link the first time (like wikipedia :))
